File: p.89/Cpp/solution.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <unordered_map>
#include <vector>
#include <filesystem>

int roman_to_int(const std::string& roman) {
  const std::unordered_map<char, int> roman_values = {
    {'I', 1}, {'V', 5}, {'X', 10}, {'L', 50},
    {'C', 100}, {'D', 500}, {'M', 1000}
  };
  int total = 0;
  int prev_value = std::numeric_limits<int>::max();

  for (char c : roman) {
    int value = roman_values.at(c);
    if (value > prev_value) {
      total += value - 2 * prev_value;
    } else {
      total += value;
    }
    prev_value = value;
  }
  return total;
}


std::string int_to_minimal_roman(int num) {
  const std::vector<std::pair<int, std::string>> mappings = {
    {1000, "M"}, {900, "CM"}, {500, "D"}, {400, "CD"},
    {100, "C"}, {90, "XC"}, {50, "L"}, {40, "XL"},
    {10, "X"}, {9, "IX"}, {5, "V"}, {4, "IV"}, {1, "I"}
  };
  std::string result;

  for (const auto& [value, numeral] : mappings) {
    while (num >= value) {
      result += numeral;
      num -= value;
    }
  }

  return result;
}

int calculate_savings(const std::string& filename) {
  std::ifstream file(filename);
  if (!file.is_open()) {
    throw std::runtime_error("Could not open file");
  }

  int savings = 0;
  std::string line;
  while (std::getline(file, line)) {
    int original_length = line.size();
    int minimal_length = int_to_minimal_roman(roman_to_int(line)).size();
    savings += original_length - minimal_length;
  }

  return savings;
}

int main() {
  const std::string filename = "roman.txt";
  try {
    int savings = calculate_savings(filename);
    std::cout << "Total characters saved: " << savings << std::endl;
  } catch (const std::exception& e) {
    std::cerr << "Error: " << e.what() << '\n';
    return 1;
  }
  return 0;
}

File: p.89/Ruby/solution.rb
def roman_to_int(roman)
  roman_values = {
    'I' => 1, 'V' => 5, 'X' => 10, 'L' => 50,
    'C' => 100, 'D' => 500, 'M' => 1000
  }
  total = 0
  prev_value = Float::INFINITY

  roman.each_char do |char|
    value = roman_values[char]
    if value > prev_value
      total += value - 2 * prev_value
    else
      total += value
    end
    prev_value = value
  end

  total
end

def int_to_minimal_roman(num)
  mappings = [
    [1000, 'M'], [900, 'CM'], [500, 'D'], [400, 'CD'],
    [100, 'C'], [90, 'XC'], [50, 'L'], [40, 'XL'],
    [10, 'X'], [9, 'IX'], [5, 'V'], [4, 'IV'], [1, 'I']
  ]

  result = ''

  mappings.each do |value, numeral|
    while num >= value
      result += numeral
      num -= value
    end
  end

  result
end


def calculate_savings(filename)
  savings = 0

  File.open(filename).each_line do |line|
    roman = line.strip
    original_length = roman.length
    minimal_length = int_to_minimal_roman(roman_to_int(roman)).length
    savings += original_length - minimal_length
  end

  savings
end

filename = 'roman.txt'
savings = calculate_savings(filename)
puts "Total characters saved: #{savings}"
